File: py_ss/pendulum_with_pi.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import matplotlib.pyplot as plt
import numpy as np

import blocks
from blocks import N as N

class Pendulum(blocks.Submodel):
    def __init__(self, iport, oport):
        super().__init__('pendulum', [iport], [oport])

        # nodes
        tau  = N(iport)
        dphi = N('dphi')
        phi  = N(oport)
        m = N('m')
        l = N('l')
        g = N('g')

        # blocks
        with self:
            blocks.MulDiv('tau/ml2', operations='*///', iports=[tau, m, l, l])
            blocks.AddSub('err', operations='+-', iports=['-', -1])
            blocks.Integrator('dphi', oport=dphi)
            blocks.Integrator('phi', iport=dphi, oport=phi)
            blocks.Function('sin(phi)', act_func=lambda t, x: np.sin(x), iport=phi)
            blocks.MulDiv('g/l', operations='**/', iports=['-', g, l], oport=-1)

class PI(blocks.Submodel):
    def __init__(self, Kp, Ki, iport, oport, x0=0.0):
        super().__init__('PI', [iport], [oport])

        # nodes
        x = N(iport)

        # blocks
        with self:
            blocks.Gain('Kp', k=Kp, iport=x, oport=-1)
            blocks.Integrator('ix', iport=x, x0=x0)
            blocks.Gain('Ki', k=Ki)
            blocks.AddSub('', operations='++', iports=[-1, '-'], oport=N(oport))

class SSModel(blocks.MainModel):
    def __init__(self):
        super().__init__('')

        # nodes
        phi = N('phi')
        tau = N('tau')
        err = N('err')

        # blocks
        with self:
            blocks.AddSub('',
                          operations='+-',
                          iports=[N('des_phi'), phi],
                          oport=err)
            PI(Kp=40.0, Ki=20.0, iport=err, oport=tau)
            Pendulum(iport=tau, oport=phi)

def main():
    parameters = {
        'm': 0.2,
        'l': 0.1,
        'g': 9.81,
        'des_phi': np.pi/4,
        }

    history = SSModel().run(parameters=parameters, t_end=5.0)

    print(history.keys())
    T = history['t']
    plt.figure()
    plt.subplot(3, 1, 1); plt.plot(T, history['phi'])
    plt.ylabel('phi')
    plt.subplot(3, 1, 2); plt.plot(T, history['dphi'])
    plt.ylabel('dphi')
    plt.subplot(3, 1, 3); plt.plot(T, history['tau'])
    plt.xlabel('t'); plt.ylabel('tau')
    plt.show()

if __name__ == '__main__':
    main()

File: cpp_ss/sandbox/pendulum_with_pi.cpp

#include <iostream>
#include <math.h>
#include <vector>

#include "blocks.hpp"
#include "helper.hpp"
#include "solver.hpp"
#include "gp-ios.hpp"

using namespace blocks;

class Pendulum : public Submodel
{
public:
    Pendulum(const Node& tau, const Node& phi) : Submodel("pendulum", {tau}, {phi})
    {
        // nodes
        auto dphi  = "dphi";
        auto m     = "m";
        auto g     = "g";
        auto l     = "l";

        // blocks
        enter();
        {
            new MulDiv("tau/ml2", "*///", {tau, m, l, l});
            new AddSub("err", "+-", {"", -1});
            new Integrator("dphi", "", dphi);
            new Integrator("phi", dphi, phi);
            // new Function("sin(phi)",
            //     [](double t, const Value& x) -> Value
            //     {
            //         return x.sin();
            //     }, phi);
            new Sin("sin(phi)", {phi});
            new MulDiv("g/l", "**/", {"", g, l}, -1);
        }
        exit();
    }
};

class PI : public Submodel
{
public:
    PI(double Kp, double Ki, Node& iport, Node& oport, double x0=0.0) :
        Submodel("PI", iport, oport)
    {
        // nodes
        auto& x = iport;

        // blocks
        enter();
        {
            new Gain("Kp", Kp, x, {-1});
            new Integrator("ix", x, "", x0);
            new Gain("Ki", Ki);
            new AddSub("", "++", {-1, ""}, oport);
        }
        exit();
    }
};

class SSModel : public Submodel
{
public:
    SSModel() : Submodel("pendulum_with_PI")
    {
        // nodes
        Node phi("phi");
        Node tau("tau");
        Node err("err");

        // blocks
        enter();
        {
            new AddSub("", "+-", {"des_phi", phi}, err);
            new PI(40.0, 20.0, err, tau);
            new Pendulum(tau, phi);
        }
        exit();
    }
};

int main()
{
    NodeValues parameters = {
        {      "m", 0.2   },
        {      "l", 0.1   },
        {      "g", 9.81  },
        {"des_phi", M_PI_4},
        };

    auto model = SSModel();
    auto history = run(model,
        [](uint k, double& t) -> bool
        {
            return arange(k, t, 0, 5, 0.01);
        },
        nullptr, parameters, rk4);

    Gnuplot gp;
	gp << "set xrange [0:500]\n";
    gp << "set yrange [-80:80]\n";
	gp << "plot" << gp.file1d((history["phi"] * (180/M_PI)).eval()) << "with lines title 'phi',"
	    << gp.file1d(history["dphi"]) << "with lines title 'dphi',"
	    << gp.file1d(history["tau"]) << "with lines title 'tau'\n";

    return 0;
}
